pkg/sfsturbo/controllerserver.go

```diff
@@ -18,179 +18,425 @@ package sfsturbo
+	"reflect"
+	"strconv"
+
 	"github.com/chnsz/golangsdk/openstack/sfs_turbo/v1/shares"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/common"
+	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/sfsturbo/services"
+	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/utils"
+	"github.com/kubernetes-csi/csi-lib-utils/protosanitizer"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
-	"k8s.io/klog"
+	log "k8s.io/klog/v2"
 )
 
 type controllerServer struct {
 	Driver *SfsTurboDriver
 }
 
-func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
-	klog.V(2).Infof("CreateVolume called with request %v", *req)
-	if err := validateCreateVolumeRequest(req); err != nil {
-		return nil, status.Error(codes.InvalidArgument, err.Error())
-	}
-
-	client, err := cs.Driver.cloud.SFSTurboV1Client()
-    if err != nil {
-		klog.V(3).Infof("Failed to create SFS Turbo v1 client: %v", err)
-		return nil, status.Error(codes.Internal, err.Error())
-    }
-
-	requestedSize := req.GetCapacityRange().GetRequiredBytes()
-    if requestedSize == 0 {
-        // At least 500GiB
-        requestedSize = 500 * tInGiB
-    }
-
-    sizeInGiB := bytesToGiB(requestedSize)
-
-	// Creating a share
-	createOpts := shares.CreateOpts{}
-	createOpts.VpcID = cs.Driver.cloud.Vpc.Id
-	createOpts.SecurityGroupID = cs.Driver.cloud.Vpc.SecurityGroupId
-	createOpts.SubnetID = cs.Driver.cloud.Vpc.SubnetId
-	createOpts.AvailabilityZone = cs.Driver.cloud.Global.AvailabilityZone
-	// build name
-	createOpts.Name = req.GetName()
-	// build share proto
-	createOpts.ShareProto = cs.Driver.shareProto
-	createOpts.Size = sizeInGiB
-	// build type
-	createOpts.ShareType = cs.Driver.cloud.Ext.ShareProto
-	if createOpts.ShareType == "" {
-		createOpts.ShareType = "STANDARD"
-		klog.V(2).Infof("Creating params STANDARD default")
-	}
-	klog.V(2).Infof("Creating params: %v", createOpts)
-
-    share, err := createShare(client, &createOpts)
+const (
+	defaultShareProto = "NFS"
+	defaultShareType  = "STANDARD"
+	minSizeInGiB      = 500
+	maxSizeInGiB      = 32768
+	expandStepInGiB   = 100
+)
+
+// resourcemode from SC
+const (
+	Availability = "availability"
+	ShareType    = "shareType"
+)
+
+func (cs *controllerServer) CreateVolume(_ context.Context, req *csi.CreateVolumeRequest) (
+	*csi.CreateVolumeResponse, error) {
+	log.Infof("CreateVolume called with request %v", protosanitizer.StripSecrets(*req))
+	cloud := cs.Driver.cloud
+
+	name := req.GetName()
+	capacityRange := req.GetCapacityRange()
+	if err := createVolumeValidation(name, capacityRange); err != nil {
+		return nil, err
+	}
+	sizeInGiB := int(utils.RoundUpSize(capacityRange.GetRequiredBytes(), common.GbByteSize))
+	// 500 ~ 32768
+	if sizeInGiB < minSizeInGiB {
+		sizeInGiB = minSizeInGiB
+	} else if sizeInGiB > maxSizeInGiB {
```

Comment:
exceeding the maximum capacity should return an error, do not create a small one

Reply:
done